assets/js/googlesitekit/datastore/user/authentication.js

```diff
@@ -0,0 +1,235 @@
+const FINISH_FETCH_AUTHENTICATION = 'FINISH_FETCH_AUTHENTICATION';
+const CATCH_FETCH_AUTHENTICATION = 'CATCH_FETCH_AUTHENTICATION';
+const RECEIVE_AUTHENTICATION = 'RECEIVE_AUTHENTICATION';
+
+export const INITIAL_STATE = {
+	authentication: undefined,
+	isFetchingAuthentication: false,
+};
+
+export const actions = {
+	*fetchAuthentication() {
+		let response, error;
+
+		yield {
+			payload: {},
+			type: START_FETCH_AUTHENTICATION,
+		};
+
+		try {
+			response = yield {
+				payload: {},
+				type: FETCH_AUTHENTICATION,
+			};
+
+			yield actions.receiveAuthentication( response );
+
+			yield {
+				payload: {},
+				type: FINISH_FETCH_AUTHENTICATION,
+			};
+		} catch ( e ) {
+			error = e;
+			yield {
+				payload: { error },
+				type: CATCH_FETCH_AUTHENTICATION,
+			};
+		}
+
+		return { response, error };
+	},
+	/**
+	 * Stores connection info received from the REST API.
+	 *
+	 * @since 1.5.0
+	 * @private
+	 * @param {Object} authentication Authentication info from the API.
+	 * @return {Object} Redux-style action.
+	 */
+	receiveAuthentication( authentication ) {
+		invariant( authentication, 'authentication is required.' );
+
+		return {
+			payload: { authentication },
+			type: RECEIVE_AUTHENTICATION,
+		};
+	},
+};
+
+export const controls = {
+	[ FETCH_AUTHENTICATION ]: () => {
+		return API.get( 'core', 'user', 'authentication' );
+	},
+};
+
+export const reducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case START_FETCH_AUTHENTICATION: {
+			return {
+				...state,
+				isFetchingAuthentication: true,
+			};
+		}
+
+		case RECEIVE_AUTHENTICATION: {
+			const { authentication } = payload;
+
+			return {
+				...state,
+				authentication,
+			};
+		}
+
+		case FINISH_FETCH_AUTHENTICATION: {
+			return {
+				...state,
+				isFetchingAuthentication: false,
+			};
+		}
+
+		case CATCH_FETCH_AUTHENTICATION: {
+			return {
+				...state,
+				error: payload.error,
+				isFetchingAuthentication: false,
+			};
+		}
+
+		default: {
+			return { ...state };
+		}
+	}
+};
+export const selectors = {
+
+	/**
+	 * Gets the authentication info for this user.
+	 *
+	 * Returns `undefined` if the authentication info is not available/loaded.
+	 *
+	 * Returns an object with the shape when successful:
+	 * ```
+	 * {
+	 *   authenticated: <Boolean>,
+	 *   grantedScopes: <Array>,
+	 *   requiredScopes: <Array>
+	 * }
+	 * ```
+	 *
+	 * @private
+	 * @since 1.8.0
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {Object|undefined} User authentication info.
+	 */
+	getAuthentication( state ) {
+		const { authentication } = state;
+		return authentication;
+	},
+	/**
+	 * Gets the Site Kit authentication status for this user.
+	 *
+	 * Returns `true` if the user is authenticated, `false` if
+	 * not. Returns `undefined` if the authentication info is not available/loaded.
+	 *
+	 * @since 1.8.0
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {boolean|undefined} Site connection status.
```

Comment:
```suggestion
	 * @return {?boolean} User authenticated status.
```

Reply:
Is this better? I guess if that's the convention: fine, but that reads like the return type is optional. Or is anything. We don't return `true|false|void`; we return `true|false|undefined`. 🤷

Reply:
We have so far used `?` over `|undefined`, so I think we should continue to do so. Unless this is actually incorrect - then we should update all instances though (which we have a ton of).

Reply:
Actually, that was a mistake. The nullable syntax is just for parameters I think, not return type. Disregard that suggestion! 😄